File: 0918_maximum_sum_circular_subarray/max_subarray_sum_circular.go
package main

import "math"

/**
 * Time: O(n)
 * Space: O(1)
 */
func maxSubarraySumCircular(A []int) int {
	maxCurr, maxSoFar := math.MinInt16, math.MinInt16
	minCurr, minSoFar := 0, 0
	total := 0
	for _, a := range A {
		maxCurr = max(a, maxCurr+a)
		maxSoFar = max(maxSoFar, maxCurr)
		minCurr = min(a, minCurr+a)
		minSoFar = min(minSoFar, minCurr)
		total += a
	}
	if maxSoFar < 0 {
		return maxSoFar
	}
	return max(maxSoFar, total-minSoFar)
}

func max(x, y int) int {
	if x >= y {
		return x
	}
	return y
}

func min(x, y int) int {
	if x <= y {
		return x
	}
	return y
}

func main() {

}

File: 0918_maximum_sum_circular_subarray/max_subarray_sum_circular.py
from typing import List


def max_subarray_sum_circular(A: List[int]) -> int:
    """
    Time: O(n)
    Space: O(1)
    """
    max_curr = max_so_far = float("-inf")
    min_curr = min_so_far = total = 0
    for a in A:
        max_curr = max(max_curr + a, a)
        max_so_far = max(max_so_far, max_curr)
        min_curr = min(min_curr + a, a)
        min_so_far = min(min_so_far, min_curr)
        total += a
    return max(max_so_far, total - min_so_far) if max_so_far > 0 else max_so_far
